azure/Guidelines.md

```diff
@@ -747,18 +747,66 @@ The ability to retry failed requests for which a client never received a respons
 <!-- Open API Spec -->
 [OpenAPI Specification]: https://github.com/OAI/OpenAPI-Specification/blob/main/versions/2.0.md
 
-### Long Running Operations & Jobs
+### Long-Running Operations & Jobs
 
-Azure generally follows the [Microsoft REST API guidelines for Long running operations](https://github.com/microsoft/api-guidelines/blob/vNext/Guidelines.md#13-long-running-operations).  Follow the Microsoft guidelines with the modifications / extensions given here when designing / implementing long running operations.
+When the processing for an operation may take a significant amount of time to complete, it should be
+implemented as a _long-running operation (LRO)_. This allows clients to continue running while the
+operation is being processed. The client determines the outcome of the operation at some later time
+through another API call.
 
-Previous Azure guidelines specified "Azure-AsyncOperation" as the name of the response header containing the operation URL,
-while the Microsoft guidelines use the name "Operation-Location".
+:white_check_mark: **DO** implement an operation as an LRO if the 99th percentile response time is greater than 1s.
 
-:white_check_mark: **DO** support both `Azure-AsyncOperation` and `Operation-Location` HEADERS, even though they are redundant so that existing SDKs and clients will continue to operate.
+There are two basic patterns that can be used for long-running operations:
+1. Resource-based long-running operations (RELO)
+2. Stepwise long-running operations
+
+:white_check_mark: **DO** use the RELO pattern when the operation is on a resource that contains a "status" property that can be used to determine the outcome of the operation.
+
```

Comment:
I think we need to add guidance on the name of the status field and terminal values for RELO too.

Reply:
Generally, you only want to wait until the resource is in a stable state when you use RELO. You don't care which operation kicked it out of the stable state. So multiple LROs to the same resource tend to - in practice - work just fine. At least I have never gotten actual customer feedback on that pattern.

Reply:
I agree that this pattern is not appropriate if users need to know the completion status of _a specific operation_, as opposed to knowing if the desired effect of their operation has been achieved.  E.g. if a VM is "failed" and a user issues a "reboot" operation on that VM, what then want to know is when the machine is available -- not necessarily that it was _their particular operation_ that caused it to become available.

So I propose adding something like the following:
> The RELO pattern should not be used in cases where the completion status of individual operations may be important to users, as opposed to simply learning that _an_ operation of the type they requested (e.g. reboot of a VM) has successfully completed.

Reply:
How should a service handle multiple concurrent long-running PATCH operations to a single resource? My guess is that the PATCH operations get queued, so they execute serially. This is good until you throw etags into the picture. If a client does a PATCH with an if-match etag the service should do what if another PATCH is already queued up? The queued PATCH may succeed and then the if-match will fail or the previous PATCH may fail and then the if-match is successful. But, the service returned a 202 accepting the PATCH before knowing the outcome. Or, if the client needs to know if a specific PATCH succeeded or failed, then they should not PATCH the actual resource itself and should issue the PATCH to an operations collection.

Maybe we say that this situation is so rare, we don't care but it makes me feel a bit uncomfortable. If we're ignoring dealing with this, then let's be sure we discussed and decided to ignore it.